// animation-timeline-js/lib/enums/timelineCursorType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineCursorType = void 0;
var TimelineCursorType;
(function (TimelineCursorType) {
    TimelineCursorType["Alias"] = "alias";
    TimelineCursorType["AllScroll"] = "all-scroll";
    TimelineCursorType["Auto"] = "auto";
    TimelineCursorType["Cell"] = "cell";
    TimelineCursorType["ContextMenu"] = "context-menu";
    TimelineCursorType["ColResize"] = "col-resize";
    TimelineCursorType["Copy"] = "copy";
    TimelineCursorType["Crosshair"] = "crosshair";
    TimelineCursorType["Default"] = "default";
    TimelineCursorType["EResize"] = "e-resize";
    TimelineCursorType["EWResize"] = "ew-resize";
    TimelineCursorType["Grab"] = "grab";
    TimelineCursorType["Grabbing"] = "grabbing";
    TimelineCursorType["Help"] = "help";
    TimelineCursorType["Move"] = "move";
    TimelineCursorType["NResize"] = "n-resize";
    TimelineCursorType["NEResize"] = "ne-resize";
    TimelineCursorType["NESWResize"] = "nesw-resize";
    TimelineCursorType["NSResize"] = "ns-resize";
    TimelineCursorType["NWResize"] = "nw-resize";
    TimelineCursorType["NWSEResize"] = "nwse-resize";
    TimelineCursorType["NoDrop"] = "no-drop";
    TimelineCursorType["None"] = "none";
    TimelineCursorType["NotAllowed"] = "not-allowed";
    TimelineCursorType["Pointer"] = "pointer";
    TimelineCursorType["Progress"] = "progress";
    TimelineCursorType["RowResize"] = "row-resize";
    TimelineCursorType["SResize"] = "s-resize";
    TimelineCursorType["SEResize"] = "se-resize";
    TimelineCursorType["SWResize"] = "sw-resize";
    TimelineCursorType["Text"] = "text";
    TimelineCursorType["WResize"] = "w-resize";
    TimelineCursorType["Wait"] = "wait";
    TimelineCursorType["ZoomIn"] = "zoom-in";
    TimelineCursorType["ZoomOut"] = "zoom-out";
})(TimelineCursorType = exports.TimelineCursorType || (exports.TimelineCursorType = {}));
